File: cuts/shrink_string.py
# The task is to write a method that takes a string like
# 'aaaabbbccd' and returns a string like 'a4b3c2d'.
# The method should throw an error on non-alphabetic characters.

def stringShrink(text: str) -> str:
    if len(text) < 1:
        return ""
    
    start_char = text[0]
    counter = 0
    dupes = []
    for ch in text:
        if not ch.isalpha():
            raise Exception
        
        if ch == start_char:
            counter += 1
            continue

        dupes.append((start_char, counter))
        start_char = ch
        counter = 1

    dupes.append((start_char, counter))
    return ''.join([f'{dupe[0]}{dupe[1] if dupe[1] > 1 else ''}' for dupe in dupes])

File: cuts/shrink_string.cpp
#include <string>
#include <vector>

using std::invalid_argument;
using std::pair;
using std::string;
using std::to_string;
using std::vector;

class Solution {
public:
    string stringShrink(const string& s) {
        if (s.size() < 1)
            return "";
        
        int counter = 0;
        char startChar = s[0];
        vector<pair<char, int>> dupes;
        for (char ch : s) {
            if (!isalpha(ch))
                throw invalid_argument("String should contain only alphabetical characters");

            if (ch == startChar) {
                counter++;
                continue;
            }

            dupes.push_back({ch, counter});
            counter = 1;
            startChar = ch;
        }

        dupes.push_back({startChar, counter});

        string result;
        for (const auto& dupe : dupes) {
            result += dupe.first;
            if (dupe.second > 1)
                result += to_string(dupe.second);
        }

        return result;
    }
};
